File: eduquestesay/lib/data/services/enrollment_service.dart
// lib/data/services/enrollment_service.dart

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/enrollment_model.dart';

class EnrollmentService {
  final String baseUrl = "http://localhost:8099/api/enrollments";

  // Enroll a student in a course
  Future<Enrollment> enrollUser(String studentEmail, String courseId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/enroll?studentEmail=$studentEmail&courseId=$courseId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> jsonResponse = json.decode(response.body);
        return Enrollment.fromJson(jsonResponse);
      } else {
        throw Exception('Failed to enroll. Status code: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to enroll: $e');
    }
  }

  // Get all enrollments
  Future<List<Enrollment>> getAllEnrollments() async {
    try {
      final response = await http.get(
        Uri.parse(baseUrl),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonResponse = json.decode(response.body);
        return jsonResponse.map((json) => Enrollment.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load enrollments. Status code: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to load enrollments: $e');
    }
  }

  // Get enrollments by student email
  Future<List<Enrollment>> getEnrollmentsByStudent(String email) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/student/$email'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonResponse = json.decode(response.body);
        return jsonResponse.map((json) => Enrollment.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load student enrollments. Status code: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to load student enrollments: $e');
    }
  }

  // Get enrollments by course ID
  Future<List<Enrollment>> getEnrollmentsByCourse(String courseId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/course/$courseId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonResponse = json.decode(response.body);
        return jsonResponse.map((json) => Enrollment.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load course enrollments. Status code: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to load course enrollments: $e');
    }
  }

  // Withdraw from a course
  Future<void> withdraw(String studentEmail, String courseId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/withdraw?studentEmail=$studentEmail&courseId=$courseId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to withdraw. Status code: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to withdraw: $e');
    }
  }

  // Update progress
  Future<Enrollment> updateProgress(String studentEmail, String courseId, double progress) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/progress?studentEmail=$studentEmail&courseId=$courseId&progress=$progress'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> jsonResponse = json.decode(response.body);
        return Enrollment.fromJson(jsonResponse);
      } else {
        throw Exception('Failed to update progress. Status code: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to update progress: $e');
    }
  }

  // Check if student is enrolled in a course
  Future<bool> isStudentEnrolled(String studentEmail, String courseId) async {
    try {
      final enrollments = await getEnrollmentsByStudent(studentEmail);
      return enrollments.any((enrollment) => enrollment.courseId == courseId);
    } catch (e) {
      return false;
    }
  }
}

File: eduquestesay_backend/eduquestesay/src/main/java/com/eduquesteasy/services/EnrollmentService.java
package com.eduquesteasy.services;

import com.eduquesteasy.models.Course;
import com.eduquesteasy.models.Enrollment;
import com.eduquesteasy.repositories.CourseRepository;
import com.eduquesteasy.repositories.EnrollmentRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class EnrollmentService {

    private final EnrollmentRepository enrollmentRepository;
    private final CourseRepository courseRepository;

    public EnrollmentService(
            EnrollmentRepository enrollmentRepository,
            CourseRepository courseRepository
    ) {
        this.enrollmentRepository = enrollmentRepository;
        this.courseRepository = courseRepository;
    }

    /**
     * Enroll a student (by email) in a course
     */
    public Enrollment enrollUser(String studentEmail, Long courseId) {

        // Check if course exists
        Course course = courseRepository.findById(courseId)
                .orElseThrow(() -> new RuntimeException("Course not found"));

        // Check if already enrolled
        enrollmentRepository.findByStudentEmailAndCourseId(studentEmail, courseId)
                .ifPresent(e -> {
                    throw new RuntimeException("Student already enrolled in this course");
                });

        // Create new enrollment
        Enrollment enrollment = new Enrollment();
        enrollment.setStudentEmail(studentEmail);
        enrollment.setCourse(course);

        return enrollmentRepository.save(enrollment);
    }

    /**
     * Get all enrollments
     */
    public List<Enrollment> getAllEnrollments() {
        return enrollmentRepository.findAll();
    }

    /**
     * Get all enrollments for a specific student
     */
    public List<Enrollment> getEnrollmentsByStudent(String studentEmail) {
        return enrollmentRepository.findByStudentEmail(studentEmail);
    }

    /**
     * Get all enrollments for a specific course
     */
    public List<Enrollment> getEnrollmentsByCourse(Long courseId) {
        return enrollmentRepository.findByCourseId(courseId);
    }

    /**
     * Withdraw student from course
     */
    public void withdraw(String studentEmail, Long courseId) {
        Enrollment enrollment = enrollmentRepository
                .findByStudentEmailAndCourseId(studentEmail, courseId)
                .orElseThrow(() -> new RuntimeException("Enrollment not found"));

        enrollmentRepository.delete(enrollment);
    }

    /**
     * Update progress of a student in a course
     */
    public Enrollment updateProgress(String studentEmail, Long courseId, double progress) {

        if (progress < 0 || progress > 100) {
            throw new RuntimeException("Progress must be between 0 and 100");
        }

        Enrollment enrollment = enrollmentRepository
                .findByStudentEmailAndCourseId(studentEmail, courseId)
                .orElseThrow(() -> new RuntimeException("Enrollment not found"));

        enrollment.setProgress(progress);

        return enrollmentRepository.save(enrollment);
    }
}
